### PR TITLE
MM-390: Added functionality to add user to all public channel in the team at once

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**
## Summary
Added functionality to add users to all the team's public channels at once.
This can be achieved by adding `*` to _ChannelsAddedTo_ in the config.json(as described in readme)

### CODE REVIEW

File: server/welcomebot.go [hunk 1]
```diff
@@ -228,12 +228,36 @@ func (p *Plugin) processActionMessage(messageTemplate MessageTemplate, action *A
 }
 
 func (p *Plugin) joinChannel(action *Action, channelName string) {
-	if channel, err := p.API.GetChannelByName(action.Context.TeamID, channelName, false); err == nil {
-		if _, err := p.API.AddChannelMember(channel.Id, action.Context.UserID); err != nil {
-			p.API.LogError("Couldn't add user to the channel, continuing to next channel", "user_id", action.Context.UserID, "channel_id", channel.Id)
-			return
+	if channelName == "*" {
+		perPage := 100
```
Comment (Kshitij-Katiyar) [inline server/welcomebot.go]: Is there a constant present plugin wide for page and perPage ?

File: server/welcomebot.go [hunk 2]
```diff
@@ -228,12 +228,36 @@ func (p *Plugin) processActionMessage(messageTemplate MessageTemplate, action *A
...
+		page := 0
+		for {
+			channels, appErr := p.API.GetPublicChannelsForTeam(action.Context.TeamID, page, perPage)
```
Comment (Kshitij-Katiyar) [inline server/welcomebot.go]: Add a comment that we will only add the users to the public channels.

**Review by Kshitij-Katiyar — APPROVED** [review]

File: server/welcomebot.go (hunk 1 above)
Comment (raghavaggarwal2308) [reply server/welcomebot.go]: not found

File: server/welcomebot.go [hunk 3]
```diff
@@ -228,12 +233,36 @@ func (p *Plugin) processActionMessage(messageTemplate MessageTemplate, action *A
...
 
 func (p *Plugin) joinChannel(action *Action, channelName string) {
-	if channel, err := p.API.GetChannelByName(action.Context.TeamID, channelName, false); err == nil {
-		if _, err := p.API.AddChannelMember(channel.Id, action.Context.UserID); err != nil {
-			p.API.LogError("Couldn't add user to the channel, continuing to next channel", "user_id", action.Context.UserID, "channel_id", channel.Id)
-			return
+	if channelName == "*" {
+		page := defaultPage
+		for {
+			// Adding user to all the public channels when channel name in '*' (i.e. all)
```
Comment (Kshitij-Katiyar) [inline server/welcomebot.go]: Maybe?
```suggestion
			// Adding user to all the public channels when channel name is '*' (i.e. all)
```

File: server/welcomebot.go [hunk 4]
```diff
@@ -228,12 +233,36 @@ func (p *Plugin) processActionMessage(messageTemplate MessageTemplate, action *A
...
 		}
 	} else {
-		p.API.LogError("failed to get channel, continuing to the next channel", "channel_name", channelName, "user_id", action.Context.UserID)
+		if channel, err := p.client.Channel.GetByName(action.Context.TeamID, channelName, false); err == nil {
+			if _, err = p.client.Channel.AddMember(channel.Id, action.Context.UserID); err != nil {
+				p.client.Log.Error("Couldn't add user to the channel, continuing to next channel", "user_id", action.Context.UserID, "channel_id", channel.Id, "error", err.Error())
+				return
+			}
+		} else {
+			p.client.Log.Error("failed to get channel, continuing to the next channel", "channel_name", channelName, "user_id", action.Context.UserID, "error", err.Error())
```
Comment (Kshitij-Katiyar) [inline server/welcomebot.go]: either use lower case or upper case like in line 252 and 242

**Review by Kshitij-Katiyar — APPROVED** [review]

**Comment by Kshitij-Katiyar** [comment]
See if there is any change we have to do in the readme file as well ?

**Comment by Kshitij-Katiyar** [comment]
Created the fork PR on mattermost-community organisation.

**Pull request closed without merging**